persistence/jpa/src/main/java/org/infinispan/persistence/jpa/JpaStore.java

```diff
@@ -441,24 +465,19 @@ public void writeBatch(Iterable<MarshalledEntry<? extends K, ? extends V>> marsh
                mergeEntity(em, entity);
                if (metadata != null && metadata.hasBytes())
                   mergeMetadata(em, metadata);
-            }
-            txn.commit();
-            stats.addBatchWriteTxCommitted(timeService.time() - txnBegin);
-         } catch (Exception e) {
-            stats.addBatchWriteTxFailed(timeService.time() - txnBegin);
-            if (e instanceof JpaStoreException)
-               throw e;
-            throw new JpaStoreException("Exception caught in writeBatch()", e);
-         } finally {
-            if (txn != null && txn.isActive())
-               txn.rollback();
-         }
-      } finally {
-         em.close();
-      }
+            })
+            .doOnComplete(() -> {
+               stats.addBatchWriteTxCommitted(timeService.time() - txnBegin);
+               txn.commit();
+            })
+            .doOnError(e -> stats.addBatchWriteTxFailed(timeService.time() - txnBegin))
```

Comment:
@wburns I ended up not using a single and just moving the stats collection to within the batch flowable method.